include/logging/log.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef LOG_H_
+#define LOG_H_
+
+#include <logging/log_instance.h>
+#include <logging/log_core.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define LOG_LEVEL_NONE  0
+#define LOG_LEVEL_ERR   1
+#define LOG_LEVEL_WARN  2
+#define LOG_LEVEL_INFO  3
+#define LOG_LEVEL_DBG   4
+
+/** @brief Module specific log level.
+ *
+ * Must be defined before including this file to override default settings.
+ *
+ * @note This header file should only be included by source files.
+ *
+ */
+#ifndef LOG_LEVEL
+#define LOG_SRC_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+#else
+#define LOG_SRC_LEVEL LOG_LEVEL
+#endif
+
+/**
+ * @brief Writes an ERROR level message to the log.
+ *
+ * @details It's meant to report severe errors, such as those from which it's
+ * not possible to recover.
+ *
+ * @param ... A string optionally containing printk valid conversion specifier,
+ * followed by as many values as specifiers.
+ */
+#define LOG_ERR(...)    LOG_INTERNAL_MODULE(LOG_LEVEL_ERR, __VA_ARGS__)
```

Comment:
So... digging through all the levels of this API...

I **think** this is going to defeat the GCC built-in printf format string validation, right?  The compiler never sees the format literal as a string passed to a function with __atribute__((format(printf...))), because it goes into a separate linker section.

If so, that's a pretty critical loss of functionality.  That warning saves a **ton** of bugs that would otherwise never be noticed.